agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -54,10 +83,12 @@ public Function thenApply(@NonNull final Function function) {
 
   @NonNull
   private Function createFunction() {
-    if (functions.isEmpty()) {
+    if (directives.isEmpty()) {
       return IDENTITY_FUNCTION;
     }
-    return new ChainFunction(functions.toArray(new Function[functions.size()]));
+    final Object[] newDirectives = directives.toArray(new Object[directives.size()]);
```

Comment:
you could make a pool of arrays, probably one array per thread would be enough and you can keep the longest array there.